### PR TITLE
fix: no warning when editing a comment to an empty comment #3411

### LINKED ISSUE #3411: [bug] No warning when editing a comment to an empty comment
**Describe the bug**
There is no warning or error message that notifies the user that an edited comment cannot be empty.

**To Reproduce**
Steps to reproduce the behaviour:

1. Write a comment eg. on one of the research threads
2. Click on Edit comment
3. Remove all the characters and make an empty comment then save
4. It doesn't change the original comment

**Expected behaviour**
When the save button is clicked, display a warning message that a comment cannot be empty

**Screenshots**
Screenshot 2024-04-02 at 11 44 59 PM
Screenshot 2024-04-02 at 11 45 07 PM
Screenshot 2024-04-02 at 11 45 23 PM

**Additional context**
I am a newbie here and want to try fixing this bug myself first if that is okay!

### PR DESCRIPTION
**Pull request by V24039**
PR Checklist

- [ ] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

Fixes the issue #3411 with the changes a error message "Comment cannot be blank" is displayed and the save is disabled until the user enters the comment for Edit Comment.

## Git Issues

Closes #3411 

## Screenshots/Videos

_If useful, provide screenshot or capture to highlight main changes_

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of our monthly maintainers call (first Monday of the month)

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

File: packages/components/src/EditComment/EditComment.tsx [hunk 1]
```diff
@@ -13,6 +14,20 @@ export interface IProps {
 
 export const EditComment = (props: IProps) => {
   const { comment, isReply } = props
+
+  const validationSchema = object({
+    comment: string().required('Make sure this field is filled correctly'),
+  })
+
+  const required = (value: string) =>
+    value && value.trim() ? undefined : 'Comment cannot be blank'
```
Comment (codisart) [inline packages/components/src/EditComment/EditComment.tsx:23]: quibble: can be replaced by `value?.trim() ? undefined : 'Comment cannot be blank'`

File: packages/components/src/EditComment/EditComment.tsx [hunk 2]
```diff
@@ -21,51 +36,69 @@ export const EditComment = (props: IProps) => {
...
-            flexDirection: 'column',
-          }}
-          p={2}
-          onSubmit={handleSubmit}
-        >
-          <Label
-            as="label"
-            htmlFor="comment"
-            sx={{ marginBottom: '6px', fontSize: 3 }}
+      validate={async (values: any) => {
```
Comment (codisart) [inline packages/components/src/EditComment/EditComment.tsx:39]: suggestion: Maybe it should be put in its own function ? Could add intention with a good name.

Comment (V24039) [reply packages/components/src/EditComment/EditComment.tsx]: created a function validateEditedComment as mentioned.

File: packages/components/src/EditComment/EditComment.tsx (hunk 1 above)
Comment (codisart) [reply packages/components/src/EditComment/EditComment.tsx]: You're welcome

**Comment by benfurber** [comment]
Thanks @V24039! Can you think what tests could cover this fix? I'm thinking in both the component library and the e2e cypress tests.

**Comment by V24039** [comment]
@benfurber I am sorry I don't know how to write test cases, from what I know test case for input field might cover it.

Again sorry for not writing test cases.

**Comment by benfurber** [comment]
@V24039 So if you look at `EditComment.test.tsx`, the test we want to setup is what you've coded.

1. EditComment has a valid comment.
2. The text is removed.
3. User clicks save.
4. The 'Comment cannot be blank' message appears.

If you have a look at `CreateComment.test.tsx` you'll see similar tests using `fireEvent.change` and `fireEvent.click`.

Happy to give it a go?

**Comment by V24039** [comment]
@benfurber Sure I can give it try and try complete it

**Comment by benfurber** [comment]
@V24039 Nice work on this. I can see that the linter is still failing. If you run `yarn format` you should get all the issues in one go.

**Comment by V24039** [comment]
@benfurber I am sorry I tried but two of test cases are failing, if possible can some please help me with this.

**Comment by benfurber** [comment]
Hey @V24039, apologies for not getting back to you yet. I've run out of time this week but if another maintainer doesn't get back to you over the weekend, I can give you a hand next week.

**Comment by V24039** [comment]
@benfurber thank you for the response.
Sure no issues, I will also try on side to find a solution

**Comment by benfurber** [comment]
Hey @V24039 I can see from the [test fails](https://app.circleci.com/pipelines/github/ONEARMY/community-platform/9006/workflows/d0c070dd-a7d4-482d-ab47-7358cff7fcfd/jobs/78521/tests) that:

---
EditComment > disables save button when comment is empty: Received element is not disabled. I think that's because `getByText('Save')` is getting the span element on the button rather than the button itself.

To fix that, on the component find `data-cy="edit-comment-submit"`, add a line below `data-testid="edit-comment-submit"`. Then on the test use `getByTestId("edit-comment-submit")`. I think that might fix it.

---
EditComment > should dispaly error message when the comment is empty: TestingLibraryElementError: Unable to find an element with the text: Comment cannot be blank. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

The message 'Comment cannot be blank` is only going to be called on submit, so don't you need to click the button?

**Comment by V24039** [comment]
@benfurber thank you for the help
The test case for button disable is now passing

The error message the validation on textarea blur, hence no submit is required to show the error message. Still trying to figure out the solution will push the code once it passes.

**Comment by V24039** [comment]
@benfurber test cases related to edit comment are now getting passed

**Comment by benfurber** [comment]
Awesome! Thank you @V24039! What did you fancy picking up next?

**Comment by benfurber** [comment]
@all-contributors please add @V24039 for code

**Review by benfurber — APPROVED** [review]

**Comment by V24039** [comment]
> Awesome! Thank you @V24039! What did you fancy picking up next?

@benfurber  Thank you for the support and approving my PR(finally my first os contribution)

Will go through the issues and try pick-up something which I am capable of.

It will my contribution to project kamp😅😅

**Pull request merged**